kyo-prelude/shared/src/test/scala/kyo2/CoroutineTest.scala

```diff
@@ -0,0 +1,255 @@
+            yield ()
+
+        val result = Coroutine.run(effect)
+        assert(result.eval == ())
+        assert(order == List(2, 3, 1))
+    }
+
+    "interleaved pauses and spawns" in {
+        var order = List.empty[Int]
+        val effect =
+            for
+                _ <- Coroutine.pause
+                _ <- Coroutine.spawn {
+                    order = 1 :: order
+                }
+                _ <- Coroutine.pause
+                _ <- Coroutine.spawn {
+                    order = 2 :: order
+                }
+            yield ()
+
+        val result = Coroutine.run(effect)
+        assert(result.eval == ())
+        assert(order == List(2, 1))
+    }
+
+    "recursive coroutine" in {
+        def recursive(n: Int): Unit < Coroutine[Any] =
+            if n == 0 then Coroutine.pause
+            else Coroutine.spawn { recursive(n - 1) }
+
+        val effect = recursive(100000)
+        val result = Coroutine.run(effect)
+        assert(result.eval == ())
+    }
+
+    "nested coroutines" in run {
+        var order = List.empty[Int]
+        val effect =
+            for
+                _ <- Coroutine.spawn {
+                    order = 1 :: order
+                    Coroutine.spawn {
+                        order = 2 :: order
+                        Coroutine.spawn {
+                            order = 3 :: order
+                        }
+                    }
+                }
+                _ <- Coroutine.pause
+                _ <- Coroutine.spawn {
+                    order = 4 :: order
+                }
+            yield ()
+
+        val result = Coroutine.run(effect)
+        assert(result.eval == ())
+        assert(order == List(3, 4, 2, 1))
+    }
+
+    "spawn with side effects" in run {
+        var a, b, c = 0
+        val effect =
+            for
+                _ <- Coroutine.spawn { a = 1 }
+                _ <- Coroutine.spawn { b = 2 }
+                _ <- Coroutine.spawn { c = 3 }
+            yield a + b + c
+
+        val result = Coroutine.run(effect)
+        assert(result.eval == 6)
+    }
+
+    "complex control flow" in run {
+        var executed = List.empty[Int]
+        def subRoutine(n: Int): Unit < Coroutine[Any] =
+            for
+                _ <- Coroutine.pause
+                _ <- Coroutine.spawn { executed = n :: executed }
+                _ <- if n > 0 then subRoutine(n - 1) else Coroutine.pause
+            yield ()
+
+        val effect =
+            for
+                _ <- Coroutine.spawn { executed = 1 :: executed }
+                _ <- subRoutine(3)
+                _ <- Coroutine.spawn { executed = 5 :: executed }
+            yield ()
+
+        val result = Coroutine.run(effect)
+        assert(result.eval == ())
+        assert(executed == List(5, 0, 1, 2, 3, 1))
+    }
+
+    "error propagation" in run {
+        val effect =
+            for
+                _ <- Coroutine.spawn { throw new RuntimeException("Spawn error") }
```

Comment:
I think it would make sense for errors to be captured by `spawn`. Perhaps we should nest `Result/Maybe`?

Reply:
Thinking about this more, it might make sense to have `spawn`  and `spawnDaemon` - a variant which can proceed when other tasks fail.

Reply:
We're limited by `spawn` returning `Unit` in this initial version of the effect. Would `spawnDaemon` swallow exceptions? I explored this approach:

```scala
inline def spawnDaemon[A, S](inline v: => Unit < (Coroutine[S] & S)): Unit < Coroutine[Abort[Throwable] & S] =
    spawn(Abort.catching[Throwable](v))
```

but `run` would still stop the execution on the first `Abort`. I think we'd need a `run` method that collects multiple failures and aggregates them. I'd prefer to keep this initial version of the effect without explicit error handling and review it once we evolve it to allow joining spawned computations since we can do something similar to the new `Fiber` design.